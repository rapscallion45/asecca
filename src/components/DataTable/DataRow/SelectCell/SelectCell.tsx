import { FC, useCallback, memo } from 'react';
import {
  FormControl,
  MenuItem,
  Select,
  SelectChangeEvent,
} from '@mui/material';
import {
  DataTableColumnSelectValueType,
  IDataTableEditSelectCellValueCallback,
} from '../../types';
import StyledTableCell from '../StyledCellWrapper/StyledCellWrapper';

/**
 * Data Table Select Cell Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @typedef ISelectCellProps
 * @prop {string} inputId - ID of the cell input field
 * @prop {boolean} canEdit - cell is editable flag
 * @prop {string | undefined} value - cell value, can be null
 * @prop {DataTableColumnSelectValueType} options - select dropdown options
 * @prop {boolean} allowUnassigned - allow the dropdown to be unassigned
 * @prop {string} unassignedText - string to display when dropdown unassigned
 * @prop {IDataTableEditSelectCellValueCallback} submitCellValue - submit cell value
 * @prop {any} sx - cell styling overrrides
 */
interface ISelectCellProps {
  inputId: string;
  canEdit: boolean;
  value: string | undefined;
  options: DataTableColumnSelectValueType;
  allowUnassigned?: boolean;
  unassignedText?: string;
  submitCellValue?: IDataTableEditSelectCellValueCallback;
  sx?: any;
}

/**
 * Data Table Select Cell
 *
 * Select table cell component for displaying a table value as a dropdown option
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @component
 * @param {ISelectCellProps} props - component props
 * @returns {FC} - data table checkbox cell functional component
 */
const SelectCell: FC<ISelectCellProps> = (props) => {
  const {
    inputId,
    canEdit,
    value,
    options,
    allowUnassigned,
    unassignedText,
    submitCellValue,
    sx,
  } = props;

  /**
   * Callback for handling user input.
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {SelectChangeEvent} event - value change trigger event
   */
  const handleValueChange = useCallback(
    (event: SelectChangeEvent) => {
      if (submitCellValue) submitCellValue(event.target.value);
    },
    [submitCellValue]
  );

  return (
    <StyledTableCell align="left" sx={sx}>
      <FormControl variant="standard" sx={{ minWidth: 230 }}>
        <Select
          id={`${inputId}-select`}
          inputProps={{ 'aria-label': inputId }}
          value={value || ''}
          onChange={handleValueChange}
          renderValue={(selectValue: string) =>
            /* if current value is undefined, display the unassigned text */
            selectValue === undefined || selectValue === '' ? (
              <MenuItem value="">{unassignedText}</MenuItem>
            ) : (
              /* we have a value, display it */
              <MenuItem value={selectValue}>{selectValue}</MenuItem>
            )
          }
          label={inputId}
          disabled={!canEdit}
          displayEmpty={allowUnassigned}
          color="secondary"
          native={false}
        >
          {allowUnassigned && <MenuItem value="">Unassigned</MenuItem>}
          {options?.map((option) => (
            <MenuItem key={option} value={option}>
              {option}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    </StyledTableCell>
  );
};

export default memo(SelectCell);
