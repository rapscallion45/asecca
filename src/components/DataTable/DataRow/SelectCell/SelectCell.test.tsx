import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Table, TableBody, TableRow } from '@mui/material';
import SelectCell from './SelectCell';

/* test select option array */
const testOptions: Array<string> = [
  'Test Option 1',
  'Test Option 2',
  'Test Option 3',
  'Test Option 4',
];

/* test value */
const testOptionIdx: number = 1;
const testValue: string | undefined = testOptions[testOptionIdx];
let mockStateValue: string | undefined = testValue;

/* test style */
const testSx = { fontWeight: 'bold' };

/* mock callback for submit cell value */
const mockSubmitCellValueCallback = jest.fn((x: string | undefined) => {
  mockStateValue = x;
});

/**
 * Data Table Select Cell Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Data Table Select Cell', () => {
  it('Renders correctly', async () => {
    /** perform snapshot test */
    const tree = renderer
      .create(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Value', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render passed value', async () => {
      /** Arrange */
      /** Act */
      render(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert */
      expect(
        screen.getByDisplayValue(mockStateValue as string)
      ).toBeInTheDocument();
    });
  });

  describe('Editing', () => {
    /** clear mock function calls and reset test value after each test */
    afterEach(() => {
      jest.clearAllMocks();
      mockStateValue = testValue;
    });

    it('Should not be editable if prop not set', async () => {
      /** Arrange */
      const canEdit: boolean = false;
      const testUpdateValue: string = testOptions[3];

      /** Act */
      const { rerender } = render(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit={canEdit}
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - no input field, button, or clear icon should be rendered */
      expect(
        screen.getByDisplayValue(mockStateValue as string)
      ).toBeInTheDocument();
      expect(screen.getByRole('button')).toBeInTheDocument();
      expect(screen.getByRole('button')).toHaveAttribute(
        'aria-haspopup',
        'listbox'
      );

      /** Act - click on select field, update value and press enter */
      fireEvent(
        screen.getByDisplayValue(mockStateValue as string),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );
      fireEvent.change(screen.getByDisplayValue(mockStateValue as string), {
        target: { value: testUpdateValue },
      });

      /** rerender the select cell with updated props (updated value) */
      rerender(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit={canEdit}
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(screen.queryByDisplayValue(testUpdateValue)).toBeNull();
      });
    });

    it('Should be editable if prop set', async () => {
      /** Arrange */
      const testUpdateValue: string = testOptions[3];

      /** Act */
      const { rerender } = render(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
                submitCellValue={mockSubmitCellValueCallback}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - select field with test value should be rendered */
      expect(
        screen.getByDisplayValue(mockStateValue as string)
      ).toBeInTheDocument();
      expect(screen.getByRole('button')).toBeInTheDocument();
      expect(screen.getByRole('button')).toHaveAttribute(
        'aria-haspopup',
        'listbox'
      );

      /** Act - click on select field, update value and press enter */
      fireEvent(
        screen.getByDisplayValue(mockStateValue as string),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );
      fireEvent.change(screen.getByDisplayValue(mockStateValue as string), {
        target: { value: testUpdateValue },
      });

      /** rerender the select cell with updated props (updated value) */
      rerender(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(screen.getByDisplayValue(testUpdateValue)).toBeInTheDocument();
      });
    });

    it('Should be allowed to be unassigned if prop set', async () => {
      /** Arrange */
      const testUpdateValue: string = 'Unassigned';

      /** Act */
      const { rerender } = render(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
                submitCellValue={mockSubmitCellValueCallback}
                allowUnassigned
                unassignedText={testUpdateValue}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - select field with test value should be rendered */
      expect(
        screen.getByDisplayValue(mockStateValue as string)
      ).toBeInTheDocument();
      expect(screen.getByRole('button')).toBeInTheDocument();
      expect(screen.getByRole('button')).toHaveAttribute(
        'aria-haspopup',
        'listbox'
      );

      /** Act - click on select field, update value and press enter */
      fireEvent(
        screen.getByDisplayValue(mockStateValue as string),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );
      fireEvent.change(screen.getByDisplayValue(mockStateValue as string), {
        target: { value: '' },
      });

      /** rerender the select cell with updated props (updated value) */
      rerender(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(screen.getByDisplayValue('')).toBeInTheDocument();
      });
    });

    it('Should submit value if dropdown option updated', async () => {
      /** Arrange */
      const testUpdateValue: string = testOptions[3];

      /** Act */
      const { rerender } = render(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
                submitCellValue={mockSubmitCellValueCallback}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - select field with test value should be rendered */
      expect(
        screen.getByDisplayValue(mockStateValue as string)
      ).toBeInTheDocument();
      expect(screen.getByRole('button')).toBeInTheDocument();
      expect(screen.getByRole('button')).toHaveAttribute(
        'aria-haspopup',
        'listbox'
      );

      /** Act - click on select field, update value and press enter */
      fireEvent(
        screen.getByDisplayValue(mockStateValue as string),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );
      fireEvent.change(screen.getByDisplayValue(mockStateValue as string), {
        target: { value: testUpdateValue },
      });

      /** rerender the select cell with updated props (updated value) */
      rerender(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(screen.getByDisplayValue(testUpdateValue)).toBeInTheDocument();
      });

      /** Assert - submit cell callback to have been called once, with formatted value */
      expect(mockSubmitCellValueCallback.mock.calls).toHaveLength(1);
      expect(mockSubmitCellValueCallback.mock.calls[0][0]).toBe(
        testUpdateValue
      );
    });
  });

  describe('Styling', () => {
    it('Should render passed styling props', async () => {
      /** Arrange */
      /** Act */
      render(
        <Table>
          <TableBody>
            <TableRow>
              <SelectCell
                canEdit
                inputId="testrow-testkey-input"
                value={mockStateValue}
                sx={testSx}
                options={testOptions}
              />
            </TableRow>
          </TableBody>
        </Table>
      );

      /** Assert */
      expect(
        screen.getByDisplayValue(mockStateValue as string)
      ).toBeInTheDocument();
    });
  });
});
