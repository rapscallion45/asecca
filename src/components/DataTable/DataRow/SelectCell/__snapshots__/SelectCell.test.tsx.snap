// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Table Select Cell Renders correctly 1`] = `
<table
  className="MuiTable-root css-rqglhn-MuiTable-root"
  role={null}
>
  <tbody
    className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1nt70hm-MuiTableCell-root"
      >
        <div
          className="MuiFormControl-root css-kbl0ib-MuiFormControl-root"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorSecondary MuiInputBase-formControl  css-1gogwgt-MuiInputBase-root-MuiInput-root-MuiSelect-root"
            onClick={[Function]}
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-label="testrow-testkey-input"
              aria-labelledby="testrow-testkey-input-select"
              className="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
              id="testrow-testkey-input-select"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <li
                className="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
                onBlur={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="menuitem"
                tabIndex={-1}
                value="Test Option 2"
              >
                Test Option 2
              </li>
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              disabled={false}
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value="Test Option 2"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
